doc/source/tune/tutorials/tune-trial-checkpoint.rst

```diff
@@ -0,0 +1,197 @@
+
+  .. TODO: for (1), link to tune fault tolerance guide. For (2), link to tune restore guide.
+
+.. note::
+    ``checkpoint_frequency`` and ``checkpoint_at_end`` will not work with Function API checkpointing.
+    These are configured manually with Function Trainable. For example, if you want to checkpoint every three
+    epochs, you can do so through:
+
+    .. literalinclude:: /tune/doc_code/trial_checkpoint.py
+        :language: python
+        :start-after: __function_api_checkpointing_periodic_start__
+        :end-before: __function_api_checkpointing_periodic_end__
+
+
+See :ref:`here for more information on creating checkpoints <air-checkpoint-ref>`.
+If using framework-specific trainers from Ray AIR, see :ref:`here <air-trainer-ref>` for
+references to framework-specific checkpoints such as `TensorflowCheckpoint`.
+
+
+.. _tune-class-trainable-checkpointing:
+
+Class API Checkpointing
+-----------------------
+
+You can also implement checkpoint/restore using the Trainable Class API:
+
+.. literalinclude:: /tune/doc_code/trial_checkpoint.py
+    :language: python
+    :start-after: __class_api_checkpointing_start__
+    :end-before: __class_api_checkpointing_end__
+
+You can checkpoint with three different mechanisms: manually, periodically, and at termination.
+
+Manual Checkpointing
+~~~~~~~~~~~~~~~~~~~~
+
+A custom Trainable can manually trigger checkpointing by returning ``should_checkpoint: True``
+(or ``tune.result.SHOULD_CHECKPOINT: True``) in the result dictionary of `step`.
+This can be especially helpful in spot instances:
+
+.. literalinclude:: /tune/doc_code/trial_checkpoint.py
+    :language: python
+    :start-after: __class_api_manual_checkpointing_start__
+    :end-before: __class_api_manual_checkpointing_end__
+
+In the above example, if ``detect_instance_preemption`` returns True, manual checkpointing can be triggered.
+
+
+Periodic Checkpointing
+~~~~~~~~~~~~~~~~~~~~~~
+
+Periodic checkpointing can be used to provide fault-tolerance for experiments.
+This can be enabled by setting ``checkpoint_frequency=<int>`` and ``max_failures=<int>`` to checkpoint trials
+every *N* iterations and recover from up to *M* crashes per trial, e.g.:
+
+.. literalinclude:: /tune/doc_code/trial_checkpoint.py
+    :language: python
+    :start-after: __class_api_periodic_checkpointing_start__
+    :end-before: __class_api_periodic_checkpointing_end__
+
+
+Checkpointing at Termination
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The checkpoint_frequency may not coincide with the exact end of an experiment.
+If you want a checkpoint to be created at the end of a trial, you can additionally set the ``checkpoint_at_end=True``:
+
+.. literalinclude:: /tune/doc_code/trial_checkpoint.py
+    :language: python
+    :start-after: __class_api_end_checkpointing_start__
+    :end-before: __class_api_end_checkpointing_end__
+
+
+Configurations
+--------------
+Checkpointing can be configured through :class:`CheckpointConfig <ray.air.config.CheckpointConfig>`.
```

Comment:
Add a quick example:

For example, the example below will save only the best checkpoint per trial, as determined by largest custom `score` metric that gets reported:

```python
from ray import tune
from ray.air import Checkpoint, CheckpointConfig, RunConfig, session


def train_fn(config):
    for i in range(1, config["num_epochs"] + 1):
        metrics = {"score": i}
        session.report(metrics, checkpoint=Checkpoint.from_dict(metrics))

tuner = tune.Tuner(
    train_fn,
    param_space={"num_epochs": 10},
    run_config=RunConfig(
        checkpoint_config=CheckpointConfig(
            num_to_keep=1,
            checkpoint_score_attribute="score",
            checkpoint_score_order="max"
        )
    ),
)
```